### PR TITLE
Automate C134719, add TabBar Browser Object Model

### PR DESCRIPTION
**Pull request by ben-c-at-moz**
# Description

Automate the linked test and Browser Object, add support to BasePage and conftest

## Bugzilla bug ID

**ID:** 1895902
**Link:** https://bugzilla.mozilla.org/show_bug.cgi?id=1895902

## Type of change

Please delete options that are not relevant.

- [x] New Test
- [x] New POM
- [x] Other Changes (Please specify)

BasePage:

* Add ActionChains attr
* Add "instawait" attr, a zero-timeout wait object
* Rewrite `loaded()` because it was not actually checking for loaded 
* Fix the element cache -- check for stale, use index for element dict key (above may have been merged in Sarina's test)
* Attempt to deflake the credential test suite, but that's another fish to fry

main conftest:

* Add screenshot method for help in debugging CI
* Add CI cli option and fixture for easier CI in future
* Add artifacts location fixture to help with above
* Add stubs for future browser logging fixtures
* Update tasks for TC

# How does this resolve / make progress on that bug?

Implements BOM for TabBar, fully automates mute/unmute test

# Screenshots / Explanations

Open a tab, play a video. Hit the mute button and expect the status to be "muted", hit the unmute button and expect the status to be "playing". Test passes on local machine (MacOS), but fails in CI due to ongoing issues with the audio path. Research is underway. Marked as xfail until such time. Screenshot shows a pass "xpass" on local.

Screenshot 2024-05-14 at 09 57 48

# Comments / Concerns

May want to dive into whether "muted"/"playing" statuses might be visible at incorrect times. May want to check on BasePage `loaded()` runtime. May want to look into other methods of caching.

### CODE REVIEW

File: modules/browser_object_tabbar.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        """Return a tab root based on either a title or an index"""
+        with self.driver.context(self.driver.CONTEXT_CHROME):
+            if isinstance(identifier, int):
+                tab = self.get_element("tab-by-index", str(identifier))
+            elif isinstance(identifier, str):
+                tab = self.get_element("tab-by-title", identifier)
+            else:
+                # if we get an unexpected type, we shouldn't assume that the user wants sys exit
+                # but we have to cause problems for them nonetheless
+                assert False
```
Comment (pdehaan) [inline modules/browser_object_tabbar.py:54]: Q: What will this do; fail the test? Does it give some useful error message, or just a failed assert and sad trombone midi?

File: taskcluster/docker/linux/Dockerfile [hunk 2]
```diff
@@ -29,6 +29,8 @@ ENV SHELL=/bin/bash \
     HOME=/builds/worker \
     PATH=/builds/worker/.local/bin:/root/.local/bin:$PATH
 
+ENV MOZ_LOG=HTMLMediaElement:4,HTMLMediaElementEvents:4,cubeb:5,PlatformDecoderModule:5,AudioSink:5,AudioSinkWrapper:5,MediaDecoderStateMachine:4,MediaDecoder:4,MediaFormatReader:5,GMP:5,EME:5
```
Comment (pdehaan) [inline taskcluster/docker/linux/Dockerfile:32]: #YOLO

(but also, maybe a comment w/ link to this arcane magic of CPP logging -- or whatever the very specific flags are?)

File: tests/tabs/conftest.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+@pytest.fixture()
+def set_prefs():
+    """Set prefs"""
+    return [("browser.tabs.delayHidingAudioPlayingIconMS", "200")]
+
+
+@pytest.fixture()
+def video_url():
+    return "https://www.youtube.com/watch?v=mAia0v3ojzw"
```
Comment (pdehaan) [inline tests/tabs/conftest.py:17]: 🤯 totally expected to be RickRoll'ed.

File: conftest.py [hunk 4]
```diff
@@ -50,16 +57,23 @@ def opt_implicit_timeout(request):
 
 
 @pytest.fixture()
-def fx_executable(request):
+def ci(request):
```
[the remark this answers is not in the document]
Comment (ben-c-at-moz) [reply conftest.py:60]: hey that is fair, maybe I should revisit this naming

File: taskcluster/docker/linux/Dockerfile (hunk 2 above)
Comment (ben-c-at-moz) [reply taskcluster/docker/linux/Dockerfile:32]: Oh no I didn't mean to leave this in there!

**Review by jrbenny35 — APPROVED** [review]
I didn't see anything that stood out, looking good!

File: modules/browser_object_tabbar.py (hunk 1 above)
Comment (jrbenny35) [reply modules/browser_object_tabbar.py:54]: It should fail the test but I see what you're getting at. These methods almost need their own tests, maybe even their own library... (foxpuppet)

File: conftest.py [hunk 5]
```diff
@@ -107,18 +121,32 @@ def driver(
...
     """
     options = Options()
-    if opt_headless:
+    if opt_headless or ci:
         options.add_argument("--headless")
     options.binary_location = fx_executable
     for opt, value in set_prefs:
         options.set_preference(opt, value)
     driver = webdriver.Firefox(options=options)
+    driver.set_window_size(1152, 864)
```
[the remark this answers is not in the document]
Comment (ben-c-at-moz) [reply conftest.py:130]: yep, good call

File: collect_executables.sh [hunk 6]
```diff
@@ -5,6 +5,7 @@
 
 ## Determine OS and arch
 UNAME_A=$(uname -a)
+echo "$UNAME_A"
```
[the remark this answers is not in the document]
Comment (ben-c-at-moz) [reply collect_executables.sh:8]: i'll clean up this script...i think i'm going to end up rewriting it in a future pr anyway, though

File: collect_executables.sh [hunk 7]
```diff
@@ -81,8 +82,11 @@ FX_LOC=$(echo "$FX_LINK_HTML" | awk -F '"' '{print $2}')
 curl -O "$FX_LOC"
 
 mv geckodriver*.tar.gz geckodriver.tar.gz
+tar -xvzf geckodriver.tar.gz
 
 if [[ $SYSTEM_NAME == "linux" ]]
 then
+    ls firefox*.tar.bz2
```
[the remark this answers is not in the document]
Comment (ben-c-at-moz) [reply collect_executables.sh:89]: will clean up

File: conftest.py [hunk 8]
```diff
@@ -107,18 +121,32 @@ def driver(
...
     yield driver
 
     driver.quit()
 
 
+@pytest.fixture()
+def screenshot(driver: webdriver.Firefox, ci: bool):
+    def _screenshot(filename):
+        if not filename.endswith(".png"):
+            filename = filename + ".png"
```
[the remark this answers is not in the document]
Comment (ben-c-at-moz) [reply conftest.py:141]: right now this is only for debugging, so I'm not super worried about escaping it...

File: modules/browser_object_tabbar.py (hunk 1 above)
Comment (ben-c-at-moz) [reply modules/browser_object_tabbar.py:54]: I would absolutely love to have a convo in future about how to implement foxpuppet here. I looked into it earlier in the repo's lifespan, but it did look kind of outdated. Maybe I just didn't understand it though!

File: tests/tabs/conftest.py (hunk 3 above)
Comment (ben-c-at-moz) [reply tests/tabs/conftest.py:17]: I would never! 👼

**Pull request merged**